Fixed #23804 -- Added RasterField for PostGIS

yellowcap opened the pull request:
- I think this  be for PostGIS >= 2.0 only, but I am not sure how to check for that
- Lacks documentation since its a first draft
- Very basic testing, needs more complete tests.

Refs #23804.

timgraham commented:
If it helps, we might consider dropping support for PostGIS 1.5 as its initial release was Feb. 2010.

yellowcap commented:
Rasters were only added to PostGIS in version 2.0, so any raster handling in previous versions would have to be through a regular binary field or so and would not support any spatial querying that could be implemented down the line for proper raster fields.

So for rasters, dropping support for PostGIS 1.5 would make sense.

timgraham commented:
PR to drop PostGIS 1.5 support: https://github.com/django/django/pull/4345

timgraham commented:
Claude believes we should keep PostGIS 1.5 support for 1.9, so you'll need to skip these tests there.

yellowcap commented:
Sounds good to me. Is there an example somewhere in the codebase for a db version check?

timgraham commented:
Yes, you can look at my proposed PR to drop PostGIS 1.5 or grep for "ops.spatial_version". By the way, I did [email the geodjango mailing list](https://groups.google.com/d/topic/geodjango/E2ZWIKeq24c/discussion) to ask for objections to dropping PostGIS 1.5 support in 1.9, so maybe that decision isn't final.

auvipy commented:
dropping support for postgis less then 2.0 in 1.9 is fine

yellowcap commented:
Is there a way of limiting the import of test models based on the db backend?

The build keeps failing for `mysql_gis` because it tries to create the model table, see the end of the [build log here](http://djangoci.com/job/pull-requests-trusty/942/database=mysql_gis,label=trusty-pr,python=python2.7/console).

I already wrapped the model defininition in a `HAS_POSTGIS_2` statement, but that does not seem to have an effect.

timgraham commented on tests/gis_tests/gdal_tests/models.py:
```diff
@@ -0,0 +1,16 @@
+from django.db import connection
+
+HAS_POSTGIS_2 = (
+    hasattr(connection.ops, 'postgis') and
```
is there a reason for using vendor checks instead of the `supports_raster` feature flag?

yellowcap replied:
There was one, but not anymore. I updated the test to use the feature flag  and moved the import of the test model into the test. That solved the problem. Thanks!

yellowcap commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -368,3 +369,40 @@ class ExtentField(GeoSelectFormatMixin, Field):
+    """
+
+    description = "Raster Field"
+    geom_type = 'RASTER'
```
This is not ideal naming, but it leverages the `geom_type` attribute checks in  the operations module, so renaming this to be more general (to include raster) would require updating the name in the geometry related code as well.

yellowcap commented on django/contrib/gis/db/models/proxy.py:
```diff
@@ -52,7 +52,9 @@ def __set__(self, obj, value):
         # The geometry type must match that of the field -- unless the
         # general GeometryField is used.
-        if isinstance(value, self._klass) and (str(value.geom_type).upper() == gtype or gtype == 'GEOMETRY'):
+        if gtype == 'RASTER':
```
Also here, the class is called `GeometryProxy` but can be used for raster instantiation as well. Renaming would require deeper change in the codebase. Not sure what the better solution is, renaming or using it as is.

yellowcap commented:
Is there a way to prevent model loading for tests apps based on database features?

Somehow the `@skipUnlessDBFeature('supports_raster')` flag is active only for non-spatial backends . For `spatiallite` and `mysql_gis`, the tests fail because of the model import for the RasterField. I have placed the `import RasterField` command into the tests and added the skip flag, but it still imports the models.

timgraham commented:
Maybe we could wait until https://github.com/django/django/pull/4452 is completed.

yellowcap commented:
Yes, that makes sense. In the meantime, I am curious to get some feedback on the RasterField logic to know if I am on the right track here.

claudep commented:
Daniel, it's time to get back at your patch. We dropped PostGIS 1.5 support in master, so one less thing to care for. There is now also option flags to ease test skipping (https://docs.djangoproject.com/en/dev/ref/models/options/#required-db-features).
Could you please rebase your patch on current master?

yellowcap commented:
I rebased on current master and made an update on the patch to use the required-db-features flag on the test models.

yellowcap commented on django/contrib/gis/db/models/__init__.py:
```diff
@@ -11,4 +11,4 @@
     GeometryField, PointField, LineStringField, PolygonField,
     MultiPointField, MultiLineStringField, MultiPolygonField,
-    GeometryCollectionField)
+    GeometryCollectionField, RasterField)
```
Not sure if this import should be conditional to the `supports_raster` db feature. Alternatively, the `RasterField.__init__` function could check for the db feature and give an error if instantiated with the wrong backend. Currently this would probably give some error from the DB that might not be explicit enough.

claudep replied:
If possible, no conditional imports.

yellowcap replied:
How about checking for raster support on initialization of the field?

https://github.com/django/django/pull/4339/files#diff-466a45d312b8b1061be8209a47db2e40R169

yellowcap replied:
That was the wrong link, sorry!

https://github.com/geodesign/django/blob/raster_field/django/contrib/gis/db/models/fields.py#L406

With that, the dummy field can be used similarly to the GeometryField when imported in the tests:

https://github.com/django/django/pull/4339/files#diff-da8fd5c3308a451e84113f8a4a810457

claudep replied:
Indeed, the idea is to not fail at import time as much as possible. However, it's not really feasible in `__init__`, because `djang.db.connection` represents only the `default` connection, and you don't know at this stage for which connection the field is created (it might be that the `postgis` database is not the default one). So you have to delay the check to a later stage, in a method which receives a `connection` object.

claudep commented on django/contrib/gis/db/backends/postgis/schema.py:
```diff
@@ -16,8 +17,8 @@ def geo_quote_name(self, name):
-        from django.contrib.gis.db.models.fields import GeometryField
-        if not isinstance(field, GeometryField):
+        from django.contrib.gis.db.models.fields import GeometryField, RasterField
+        if not isinstance(field, (GeometryField, RasterField)):
```
Could we use BaseSpatialField here?

yellowcap commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -371,3 +392,30 @@ class ExtentField(GeoSelectFormatMixin, Field):
+    def from_db_value(self, value, expression, connection, context):
+        return connection.ops.parse_raster(value)
+
+    def get_db_prep_value(self, value, connection, prepared=False):
```
Moved the ImproperlyConfigured error into the `get_db_prep_value` function, and added tests for the exceptin in the [migrations tests](https://github.com/geodesign/django/blob/c9a5ce90aba708ea65b1263bb95fccb3e88cda75/tests/gis_tests/gis_migrations/test_operations.py#L126) and the [field tests](https://github.com/geodesign/django/blob/c9a5ce90aba708ea65b1263bb95fccb3e88cda75/tests/gis_tests/rasterapp/test_rasterfield.py#L85).

claudep commented on tests/gis_tests/rasterapp/test_rasterfield.py:
```diff
@@ -0,0 +1,100 @@
+             [10.0, 11.0, 12.0, 13.0, 14.0],
+             [15.0, 16.0, 17.0, 18.0, 19.0],
+             [20.0, 21.0, 22.0, 23.0, 24.0]],
+            band
```
This assertion fails currently when numpy is not available, as the `data()` method is returning a flat list in this case. Is this expected behavior, or should `data()` emulate the list reshape as a list of lists?

yellowcap replied:
Thanks for pointing this out. This is expected behaviour, I think raster data is usually stored as straight arrays, so the flat array is the "original" format. That is why in my opinion the list should be flat and leave it to the user to convert into a nested list if that is desired. In numpy, specifying the shape is not a costly operation (correct me if I am wrong here), so the shape can be specified for convenience, as it prints nicely and its easier to access specific pixel values.

So I added the `flatten()` argument to the numpy case for the test to work without numpy as well.

yellowcap commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -371,3 +392,37 @@ class ExtentField(GeoSelectFormatMixin, Field):
+    description = "Raster Field"
+    geom_type = 'RASTER'
+
+    def _check_connection(self, connection):
```
I added checks for the connection feature to `db_type` (which is called on table creation I think) and the `get_db_prep_value` (which is called on adding a field through migrations). There are tests for this exception in the [migrations tests](https://github.com/geodesign/django/blob/c9a5ce90aba708ea65b1263bb95fccb3e88cda75/tests/gis_tests/gis_migrations/test_operations.py#L126) and the [field tests](https://github.com/geodesign/django/blob/c9a5ce90aba708ea65b1263bb95fccb3e88cda75/tests/gis_tests/rasterapp/test_rasterfield.py#L85). With this, all the tests pass and the creation of raster fields on non-raster databases will raise an exception.

yellowcap commented:
As a side note: I opened a [related pull request](https://github.com/django/django/pull/4694) to add `Transform` and `Warp` functions to the `GDALRaster` api. With this patch I was able to create a branch of Django which has all the requirements for being used with an [adopted branch](https://github.com/geodesign/django-raster/tree/django_1_9) of the django-raster package. The adopted django-raster branch does no longer depend on python-gdal. I am testing the full django based raster workflow in one of my applied projects and it works just fine!

claudep commented:
@timgraham Time for your super-duper final review?

timgraham commented on django/contrib/gis/db/backends/postgis/const.py:
```diff
@@ -0,0 +1,43 @@
+# Struct pack structure for raster header, the raster header has the
+# following structure:
+#
+# Endianness, PostGIS raster version, number of bands, scale, origin
```
origin,

timgraham commented on django/contrib/gis/db/backends/postgis/const.py:
```diff
@@ -0,0 +1,43 @@
+# Endianness, PostGIS raster version, number of bands, scale, origin
+# skew, srid, width, and height.
+#
+# Scale, origin and skew have x and y values. PostGIS currently uses
```
origin,

timgraham commented on django/contrib/gis/db/backends/postgis/const.py:
```diff
@@ -0,0 +1,43 @@
+    'l': 4,  # Long
+    'L': 4,  # Unsigned Long
+    'f': 4,  # Float
+    'd': 8   # Double
```
8,

timgraham commented on django/contrib/gis/db/backends/postgis/operations.py:
```diff
@@ -266,10 +269,21 @@ def get_geom_placeholder(self, f, value, compiler):
+            value_srid = value.srid
+
+        # Adding Transform() to the SQL placeholder if the value srid
+        # not equal to the field srid.
```
is not

timgraham commented on django/contrib/gis/db/backends/postgis/operations.py:
```diff
@@ -353,3 +367,11 @@ def geometry_columns(self):
     def spatial_ref_sys(self):
         return PostGISSpatialRefSys
+
+    # Routines to convert between PostGIS rasters and dicts that are
```
Routines -> Methods

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+
+def pack(structure, data):
+    """
+    Packs data into hex string with little endian format.
```
Lastest decision is to use pep8 style docstrings for new code. e.g. "Packs -> Pack"

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+
+def unpack(structure, data):
+    """
+    Unpacks little endian hexlified binary string into python list.
```
python -> a

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+
+def from_pgraster(data):
+    """
+    Converts a PostGIS HEX String into a Python dictionary.
```
chop "Python"

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+    """
+    Converts a PostGIS HEX String into a Python dictionary.
+    """
+    # Return if input data is null
```
a bit obvious?

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+        pixeltypes.append(pixeltype)
+
+    # Check that all bands have the same pixeltype.
+    # This is required by GDAL, PostGIS rasters could have different pixeltypes
```
, -> .

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+    if len(set(pixeltypes)) != 1:
+        raise ValidationError("Band pixeltypes are not all equal.")
+
+    # Return the parsed raster as a dictionary
```
repetitive with method docstring

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+    # Prepare the raster header data as tuple. The first two numbers are
+    # the endianness and the PostGIS Raster Version, both are fixed by
+    # PostGIS at the moment.
+    rasterheader = (1, 0, len(rast.bands), rast.scale.x, rast.scale.y,
```
prefer hanging indent style (allows longer lines, indentation always multiple of 4):
```
rasterheader = (
    1, 0, ...
)
```

timgraham commented on django/contrib/gis/db/backends/postgis/pgraster.py:
```diff
@@ -0,0 +1,161 @@
+    if rast is None or rast == '':
+        return
+
+    # Prepare the raster header data as tuple. The first two numbers are
```
as a tuple

timgraham commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -371,3 +395,37 @@ class ExtentField(GeoSelectFormatMixin, Field):
+
+class RasterField(BaseSpatialField):
+    """
+    Raster field for GeoDjango.
```
Not sure that description adds much value.

timgraham commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -371,3 +395,37 @@ class ExtentField(GeoSelectFormatMixin, Field):
+    Raster field for GeoDjango.
+    """
+
+    description = "Raster Field"
```
should this be translated?  `_("Raster field")`?

timgraham commented on django/contrib/gis/db/models/fields.py:
```diff
@@ -371,3 +395,37 @@ class ExtentField(GeoSelectFormatMixin, Field):
+    def contribute_to_class(self, cls, name, **kwargs):
+        super(RasterField, self).contribute_to_class(cls, name, **kwargs)
+
+        # Setup for lazy-instantiated Raster object.
```
say why the object should be lazily instantiated?

timgraham commented on django/contrib/gis/db/models/proxy.py:
```diff
@@ -1,66 +1,71 @@
-        if isinstance(value, self._klass) and (str(value.geom_type).upper() == gtype or gtype == 'GEOMETRY'):
-            # Assigning the SRID to the geometry.
+        if gtype == 'RASTER' and (value is None or isinstance(value, six.string_types + (dict, self._klass))):
+            # For raster fields, assure input is None or a string, dict or raster instance.
```
dict,

timgraham commented on tests/gis_tests/gis_migrations/test_operations.py:
```diff
@@ -41,15 +45,24 @@ def apply_operations(self, app_label, project_state, operations):
-    def set_up_test_model(self):
+    def set_up_test_model(self, force_raster_creation=False):
+        # Add regular fields to test model
+        testfields = [
```
`test_fields` or simply `fields` if no name collision

timgraham commented on tests/gis_tests/gis_migrations/test_operations.py:
```diff
@@ -41,15 +45,24 @@ def apply_operations(self, app_label, project_state, operations):
+            ('name', models.CharField(max_length=100, unique=True)),
+        ]
+        # For GIS enabled DB add geometry field
+        if connection.features.gis_enabled:
```
not sure why this check is needed -- don't these tests only run on `gis_enabled` backends?

timgraham commented on tests/gis_tests/gis_migrations/test_operations.py:
```diff
@@ -63,87 +76,131 @@ def assertGeometryColumnsCount(self, expected_count):
+            indexes = connection.introspection.get_indexes(cursor, table)
+        self.assertIn(column, indexes)
+
+    def alter_gis_model(self, migrationclass, modelname, fieldname, blank=False, fieldclass=None):
```
migration_class, model_name, etc.

timgraham commented on tests/gis_tests/gis_migrations/test_operations.py:
```diff
@@ -63,87 +76,131 @@ def assertGeometryColumnsCount(self, expected_count):
         if HAS_GEOMETRY_COLUMNS:
             self.assertGeometryColumnsCount(2)
 
+        # Test Spatial Indices when available
```
spatial indices

timgraham commented on tests/gis_tests/gis_migrations/test_operations.py:
```diff
@@ -63,87 +76,131 @@ def assertGeometryColumnsCount(self, expected_count):
-            self.assertIn('path', indexes)
+            self.assertSpatialIndexExists('gis_neighborhood', 'path')
+
+    @skipUnlessDBFeature("supports_raster")
```
try to be consistent and use single quotes in most places (sorry existing code is inconsistent)

timgraham commented on tests/gis_tests/gis_migrations/test_operations.py:
```diff
@@ -63,87 +76,131 @@ def assertGeometryColumnsCount(self, expected_count):
+        """
+        Test creating a model with a raster field on a db without raster support.
+        """
+        self.assertRaises(ImproperlyConfigured, self.set_up_test_model, True)
```
Use the context manager version for better readability:
```
msg = '....'
with self.assertRaisesMessage(ImproperlyConfigured, msg):
    self.set_up_test_model(True)
```

timgraham commented on tests/gis_tests/rasterapp/models.py:
```diff
@@ -0,0 +1,12 @@
+
+
+class RasterModel(models.Model):
+
```
no blank line before first field

timgraham commented on tests/gis_tests/rasterapp/test_rasterfield.py:
```diff
@@ -0,0 +1,83 @@
+        """
+        Test creating a model where the RasterField has a null value.
+        """
+        rastermodel = RasterModel.objects.create(rast=None)
```
`raster_model` or even just `r` is probably fine.

timgraham commented on tests/gis_tests/rasterapp/test_rasterfield.py:
```diff
@@ -0,0 +1,83 @@
+
+        # Test raster metadata properties
+        self.assertEqual((5, 5), (rastermodel.rast.width, rastermodel.rast.height))
+        self.assertEqual([0.0, -1.0, 0.0, 0.0, 0.0, 1.0],
```
single line is fine

timgraham commented on tests/gis_tests/rasterapp/test_rasterfield.py:
```diff
@@ -0,0 +1,83 @@
+        self.assertEqual([0.0, -1.0, 0.0, 0.0, 0.0, 1.0],
+                         rastermodel.rast.geotransform)
+
+        self.assertEqual(None, rastermodel.rast.bands[0].nodata_value)
```
assertIsNone

timgraham commented:
Time to add docs?

yellowcap commented:
I added documentation for the RasterField. I hope I found all the relevant bits, I was not sure how to make clear that this is only supported for PostGIS backends, so I added it in one or two places where I found it relevant.

It would probably be good to also integrate the raster field into the GeoDjango Tutorial, but maybe that is another pull request?

claudep commented:
Thanks! I'll let Tim review the language.

timgraham commented:
Minor edits: http://dpaste.com/1YSTF0M and add release notes too?

yellowcap commented:
Integrated comments and added release notes.

timgraham commented:
please squash commits, thanks

yellowcap commented:
Corrected those glitches and squashed commits.

timgraham commented:
merged in b769bbd4f6a3cd1bcd9ebf3559ec6ea0f9b50565, thanks!

claudep commented:
Awesome, congrats Daniel!

yellowcap commented:
Feels good! A big thanks you both for all the help!

timgraham commented:
Sorry, I forgot to try running the tests on systems without the GIS dependencies installed. Daniel, can you take a look? https://code.djangoproject.com/ticket/25011

yellowcap commented:
Here is a patch for it https://github.com/django/django/pull/4905

I think the same problem will also happen on systems that have only geos installed, so I think the ``RasterField`` import should be conditional on ``HAS_GDAL`` (which is kind of obvious, but have not thought about that before).
